### PR TITLE
Add a solution design template

### PR DESCRIPTION
**Pull request by orenbm**
Having a template for solution designs (and other templates that will follow)
can help us maintain a standard for future designs.

# Notes for the reviewer
- Can you think of a section that is missing?
- Would you remove any section?
- Are the comments clear enough?

Note that the MD file will not include the comments that help the write so they don't need to be deleted, as can be seen [here](https://github.com/cyberark/conjur/blob/solution-design-template/design/templates/solution_design.md).

### CODE REVIEW

**Comment by InbalZilberman** [comment]
Hi, I like this template very much
Few things I am missing:
1. Audit section. We need to always think if this new feature has Audit implications.
2. The word options- a good design in my opinion has thought of several options and took from it the best one. Having the word option would encourage imo people to think of more than one way to solve a problem

**Comment by orenbm** [comment]
Thanks @InbalZilberman for your comments. 

> The word options- a good design in my opinion has thought of several options and took from it the best one. Having the word option would encourage imo people to think of more than one way to solve a problem

I'm not sure i follow. In the `Solution` section I say:
`If there are a few options considered for the solution, mention them and explain why the actual solution was chosen over them.`
isn't that what you want the language to be?

File: design/templates/solution_design.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+[//]: # (Add links that may be useful for the reader)
+
+## Background
+[//]: # (Give relevant background for the designed feature. What is the motivation for this solution?)
+
+## Issue description
+[//]: # (Elaborate on the issue you are writing a solution for)
+
+## Solution
+[//]: # (Elaborate on the solution you are suggesting in this page. Address the functional requirements and the non functional requirements that this solution is addressing. If there are a few options considered for the solution, mention them and explain why the actual solution was chosen over them. Add an execution plan when relevant. It doesn't have to be a full breaking of the feature, but just a recommendation to how the solution should be approached.)
```
Comment (izgeri) [inline design/templates/solution_design.md:26]:
```suggestion
[//]: # (Elaborate on the solution you are suggesting in this page. Address the functional requirements and the non functional requirements that this solution is addressing. If there are a few options considered for the solution, mention them and explain why the actual solution was chosen over them. Add an execution plan when relevant. It doesn't have to be a full breakdown of the feature, but just a recommendation to how the solution should be approached.)
```

File: design/templates/solution_design.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+### Design
+[//]: # (Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about what the expected user experience for the feature)
```
Comment (izgeri) [inline design/templates/solution_design.md:29]:
```suggestion
[//]: # (Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature)
```

File: design/templates/solution_design.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+|              |                 |
+
+### Audit 
+[//]: # (Does this solution require additional audit messages?)
+
+## Documentation
+[//]: # (Add notes on what should be documented in this solution. Elaborate on where this should be documented. If the change is in open-source projects, we may need to update the docs in github too. If it's in Conjur and/or DAP mention which products are affected by it)
+
+## Version update
+[//]: # (Does this solution require a version update? if so, add a list of the projects that should be bumped)
```
Comment (izgeri) [inline design/templates/solution_design.md:68]: this section doesn't make sense to me - in order to use any changes we make to this codebase, we must include the changes in a new version

I think what you are really asking is for other dependencies that may also need to be updated as a result of this change. wouldn't this be covered in affected components?

Maybe that section should be "considerations" and include this bullet (maybe reworded to be a bit more clear), the "affected components" bullet, and the "backwards compatibility" bullet?

File: design/templates/solution_design.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+[//]: # (Elaborate on the issue you are writing a solution for)
+
+## Solution
+[//]: # (Elaborate on the solution you are suggesting in this page. Address the functional requirements and the non functional requirements that this solution is addressing. If there are a few options considered for the solution, mention them and explain why the actual solution was chosen over them. Add an execution plan when relevant. It doesn't have to be a full breaking of the feature, but just a recommendation to how the solution should be approached.)
+
+### Design
+[//]: # (Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about what the expected user experience for the feature)
+
+### Backwards compatibility
+[//]: # (Address how you are going to handle backwards compatibility, if necessary)
```
Comment (izgeri) [inline design/templates/solution_design.md:32]: is it necessary to have a whole separate section on this?

File: design/templates/solution_design.md (hunk 1 above)
Comment (orenbm) [reply design/templates/solution_design.md:26]: updated the doc in the original commit, thanks.

File: design/templates/solution_design.md (hunk 3 above)
Comment (orenbm) [reply design/templates/solution_design.md:68]: You are correct. I removed this section as it is covered in "affected components"

File: design/templates/solution_design.md (hunk 4 above)
Comment (orenbm) [reply design/templates/solution_design.md:32]: I think it is important enough to have a sub-section under "Solution". It can be interesting to see how the solution handles the backwards-compatibility.

**Comment by orenbm** [comment]
@izgeri can you please re-review?

**Comment by InbalZilberman** [comment]
@orenbm looks good to me

File: design/templates/solution_design.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+[//]: # (Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature)
+
+### Backwards compatibility
+[//]: # (Address how you are going to handle backwards compatibility, if necessary)
+
+### Performance
+[//]: # (Elaborate on whether your solution will affect the product's performance, and how)
+
+### Affected Components
+[//]: # (Address the componentes that will be affected by your solution [Conjur, DAP, etc.])
```
Comment (izgeri) [inline design/templates/solution_design.md:38]:
```suggestion
[//]: # (Address the components that will be affected by your solution [Conjur, DAP, clients, integrations, etc.])
```

File: design/templates/solution_design.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+### Audit 
+[//]: # (Does this solution require additional audit messages?)
+
+## Documentation
+[//]: # (Add notes on what should be documented in this solution. Elaborate on where this should be documented. If the change is in open-source projects, we may need to update the docs in github too. If it's in Conjur and/or DAP mention which products are affected by it)
+
+## Open questions
+[//]: # (Add any question that is still open. It makes it easier for the reader to have the open questions accumulated here istead of them being acattered along the doc)
+
+## Breaking
```
Comment (izgeri) [inline design/templates/solution_design.md:70]:
```suggestion
## Implementation plan
```

**Comment by izgeri** [comment]
had to rebase against master to merge

**Comment by izgeri** [comment]
had to rebase against master to merge, will merge when green

**Comment by InbalZilberman** [comment]
@orenbm all my concerns were answered
